File: src/main/java/io/felipeandrade/gsw2/block/GSWMaterialBlock.java
package io.felipeandrade.gsw2.block;

import net.fabricmc.fabric.api.block.FabricBlockSettings;
import net.minecraft.block.Block;
import net.minecraft.block.Blocks;

import io.felipeandrade.gsw2.GSWItemGroup;
import io.felipeandrade.gsw2.material.GSWMaterial;

public class GSWMaterialBlock extends GSWBlock {

    public static final Settings SETTINGS_ORE = FabricBlockSettings.copy(Blocks.IRON_ORE).build();

    public static final Settings SETTINGS_METAL = FabricBlockSettings.copy(Blocks.IRON_BLOCK).build();
    public static final Settings SETTINGS_GEM = FabricBlockSettings.copy(Blocks.EMERALD_BLOCK).build();

    protected final GSWMaterial material;

    public GSWMaterialBlock(GSWMaterial material, Block.Settings settings) {
        this(material.unlocalizedName() + "_block", material, settings);
    }

    public GSWMaterialBlock(String unlocalizedName, GSWMaterial material, Settings settings) {
        super(unlocalizedName, GSWItemGroup.MATERIALS, settings);
        this.material = material;
    }
}

File: gsw-mod/src/main/kotlin/io/felipeandrade/gsw/block/GSWMaterialBlock.kt
package io.felipeandrade.gsw.block

import io.felipeandrade.gsw.GSWItemGroup
import io.felipeandrade.gsw.material.GSWMaterial
import net.fabricmc.fabric.api.`object`.builder.v1.block.FabricBlockSettings
import net.minecraft.block.Blocks

open class GSWMaterialBlock(
    val gswMaterial: GSWMaterial,
    settings: Settings,
    unlocalizedName: String = "block",
) : GSWBlock("${gswMaterial.unlocalizedName}_$unlocalizedName", GSWItemGroup.MATERIALS, settings) {

    companion object {
        val SETTINGS_ORE: Settings = FabricBlockSettings.copy(Blocks.IRON_ORE)
        val SETTINGS_DEEPSLATE_ORE: Settings = FabricBlockSettings.copy(Blocks.DEEPSLATE_IRON_ORE)
        val SETTINGS_METAL: Settings = FabricBlockSettings.copy(Blocks.IRON_BLOCK)
        val SETTINGS_GEM: Settings = FabricBlockSettings.copy(Blocks.EMERALD_BLOCK)
    }
}